README.md
![Pub Version (including pre-releases)](https://img.shields.io/pub/v/helpers_plethora_flutter?include_prereleases)
![GitHub Top Language](https://img.shields.io/github/languages/top/Marco87Developer/helpers_plethora_flutter)

An helpful variety of extensions and enumerations dependent on Flutter.

## Usage

### `DateTime` extension methods

You can determine if a date belongs to a date range:

```dart
final DateTime start = DateTime(2022, 7);
final DateTime end = DateTime(2022, 7, 31);
final DateTimeRange range = DateTimeRange(start: start, end: end);
DateTime(2022, 7, 15).belongsTo(range);
```

### `DateTimeRange` extension methods

You can determine if a date range **belongs to** another date range:

```dart
final DateTimeRange other =
    DateTimeRange(
      start: DateTime(2022),
      end: DateTime(2022, 7, 31),
    );
DateTimeRange(
  start: DateTime(2021, 9),
  end: DateTime(2022),
).belongsTo(other);
```

You can determine if a date range **overlaps**, in part or completely, with the other date range:

```dart
final DateTimeRange other =
    DateTimeRange(
      start: DateTime(2022),
      end: DateTime(2022, 7, 31),
    );
DateTimeRange(
  start: DateTime(2021, 9),
  end: DateTime(2022, 1, 2),
).overlapsWith(other);
```

### `List<DateTimeRange>` extension methods

You can get the sum of the **duration** of *all* intervals in a list *after eliminating all duplicate intervals* (which overlap some interval already present in the list).

```dart
final DateTimeRange range1 =
    DateTimeRange(start: DateTime(2022), end: DateTime(2022, 7));
final DateTimeRange range2 = DateTimeRange(
  start: DateTime(2022, 1, 25),
  end: DateTime(2022, 4, 7),
);
final DateTimeRange range3 =
    DateTimeRange(start: DateTime(2022, 4, 7), end: DateTime(2022, 7));
final DateTimeRange range4 = DateTimeRange(
  start: DateTime(2022, 5, 2),
  end: DateTime(2022, 6, 15),
);
final DateTimeRange range5 = DateTimeRange(
  start: DateTime(2022, 7, 3),
  end: DateTime(2022, 9, 30),
);
final DateTimeRange range6 =
    DateTimeRange(start: DateTime(2022), end: DateTime(2022, 7));
final DateTimeRange range7 = range5;
<DateTimeRange>[range1, range2, range3, range4, range5].duration.inDays;
```

You can get the a **list of date ranges** that **covers the entirety** of the period of the sum of all date ranges in this list, but *without repeating any overlap*:

```dart
final DateTimeRange range1 =
    DateTimeRange(start: DateTime(2022), end: DateTime(2022, 7));
final DateTimeRange range2 = DateTimeRange(
  start: DateTime(2022, 1, 25),
  end: DateTime(2022, 4, 7),
);
final DateTimeRange range3 =
    DateTimeRange(start: DateTime(2022, 4, 7), end: DateTime(2022, 7));
final DateTimeRange range4 = DateTimeRange(
  start: DateTime(2022, 5, 2),
  end: DateTime(2022, 6, 15),
);
final DateTimeRange range5 = DateTimeRange(
  start: DateTime(2022, 7, 3),
  end: DateTime(2022, 9, 30),
);
<DateTimeRange>[range1, range2, range3, range4, range5].minimize();
```

### `TimeOfDay` extension methods

You can determine if a time **is after** another (this > other):

```dart
const TimeOfDay(hour: 7, minute: 35)
    .isAfter(const TimeOfDay(hour: 7, minute: 57));
```

You can determine if a time **is before** another (this < other):

```dart
const TimeOfDay(hour: 12, minute: 20)
    .isBefore(const TimeOfDay(hour: 7, minute: 10));
```

You can get the **string representation** of a time:

```dart
const TimeOfDay(hour: 7, minute: 15).toStringHm();
```


CHANGELOG.md
## 1.0.1

Release date: 2022-10-02.

### Removed

* GitHub Dart action badge.

## 1.0.0

Release date: 2022-10-02.

* First release.


test/extensions/datetime_extension_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:helpers_plethora_flutter/src/extensions/datetime_extension.dart';

void main() {
  group('belongsTo:', () {
    test(
      'Must return whether this date belongs to the date range date range'
      ' (including extremes) or not (start ⩽ this ⩽ end).',
      () {
        final DateTime start = DateTime(2022, 7);
        final DateTime end = DateTime(2022, 7, 31);
        final DateTimeRange range = DateTimeRange(start: start, end: end);
        expect(DateTime(2022, 6, 30).belongsTo(range), isFalse);
        expect(DateTime(2022, 7).belongsTo(range), isTrue);
        expect(DateTime(2022, 7, 15).belongsTo(range), isTrue);
        expect(DateTime(2022, 7, 31).belongsTo(range), isTrue);
        expect(DateTime(2022, 8).belongsTo(range), isFalse);
      },
    );
  });
}


test/extensions/timeofday_extension_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:helpers_plethora_flutter/src/extensions/timeofday_extension.dart';

void main() {
  group('isAfter:', () {
    test('Must return if a time is after another one (>).', () {
      expect(
        const TimeOfDay(hour: 5, minute: 57)
            .isAfter(const TimeOfDay(hour: 7, minute: 12)),
        isFalse,
      );
      expect(
        const TimeOfDay(hour: 7, minute: 35)
            .isAfter(const TimeOfDay(hour: 7, minute: 57)),
        isFalse,
      );
      expect(
        const TimeOfDay(hour: 7, minute: 35)
            .isAfter(const TimeOfDay(hour: 7, minute: 35)),
        isFalse,
      );
      expect(
        const TimeOfDay(hour: 7, minute: 35)
            .isAfter(const TimeOfDay(hour: 7, minute: 34)),
        isTrue,
      );
    });
  });

  group('isBefore:', () {
    test('Must return if a time is before another one (<).', () {
      expect(
        const TimeOfDay(hour: 7, minute: 35)
            .isBefore(const TimeOfDay(hour: 7, minute: 57)),
        isTrue,
      );
      expect(
        const TimeOfDay(hour: 7, minute: 35)
            .isBefore(const TimeOfDay(hour: 7, minute: 35)),
        isFalse,
      );
      expect(
        const TimeOfDay(hour: 7, minute: 35)
            .isBefore(const TimeOfDay(hour: 7, minute: 34)),
        isFalse,
      );
      expect(
        const TimeOfDay(hour: 12, minute: 20)
            .isBefore(const TimeOfDay(hour: 7, minute: 10)),
        isFalse,
      );
    });
  });

  group('toStringHm:', () {
    test(
      'Must return the localized string representation of this time with the'
      ' hours and the minutes.',
      () {
        expect(const TimeOfDay(hour: 7, minute: 15).toStringHm(), '07:15');
        expect(const TimeOfDay(hour: 15, minute: 5).toStringHm(), '15:05');
        expect(const TimeOfDay(hour: 12, minute: 35).toStringHm(), '12:35');
      },
    );
  });
}


test/extensions/datetimerange_extension_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:helpers_plethora_flutter/src/extensions/datetimerange_extension.dart';

void main() {
  group('belongsTo:', () {
    test(
      'Must return if this date range belongs to the other one, that is, it is'
      ' completely included in it (other.start ⩽ [start] ⩽ other.start.end,'
      ' and other.start ⩽ [end] ⩽ other.end).',
      () {
        final DateTimeRange other =
            DateTimeRange(start: DateTime(2022), end: DateTime(2022, 7, 31));
        expect(
          DateTimeRange(
            start: DateTime(2021, 9),
            end: DateTime(2022),
          ).belongsTo(other),
          isFalse,
        );
        expect(
          DateTimeRange(
            start: DateTime(2021, 9),
            end: DateTime(2022, 1, 2),
          ).belongsTo(other),
          isFalse,
        );
        expect(
          DateTimeRange(
            start: DateTime(2022),
            end: DateTime(2022, 7, 31),
          ).belongsTo(other),
          isTrue,
        );
        expect(
          DateTimeRange(
            start: DateTime(2022, 1, 2),
            end: DateTime(2022, 7, 30),
          ).belongsTo(other),
          isTrue,
        );
        expect(
          DateTimeRange(
            start: DateTime(2022, 7, 30),
            end: DateTime(2022, 8),
          ).belongsTo(other),
          isFalse,
        );
      },
    );
  });

  group('overlapsWith:', () {
    test(
      'Must return if this date range overlaps, in part or completely, with the'
      ' other one (other.start ⩽ start ⩽ other.end, or other.start ⩽ end ⩽'
      ' other.end).',
      () {
        final DateTimeRange other =
            DateTimeRange(start: DateTime(2022), end: DateTime(2022, 7, 31));
        expect(
          DateTimeRange(
            start: DateTime(2021, 9),
            end: DateTime(2021, 9, 5),
          ).overlapsWith(other),
          isFalse,
        );
        expect(
          DateTimeRange(
            start: DateTime(2021, 9),
            end: DateTime(2022),
          ).overlapsWith(other),
          isTrue,
        );
        expect(
          DateTimeRange(
            start: DateTime(2021, 9),
            end: DateTime(2022, 1, 2),
          ).overlapsWith(other),
          isTrue,
        );
        expect(
          DateTimeRange(
            start: DateTime(2022),
            end: DateTime(2022, 7, 31),
          ).overlapsWith(other),
          isTrue,
        );
        expect(
          DateTimeRange(
            start: DateTime(2022, 1, 2),
            end: DateTime(2022, 7, 30),
          ).overlapsWith(other),
          isTrue,
        );
        expect(
          DateTimeRange(
            start: DateTime(2022, 7, 30),
            end: DateTime(2022, 8),
          ).overlapsWith(other),
          isTrue,
        );
        expect(
          DateTimeRange(
            start: DateTime(2022, 8),
            end: DateTime(2022, 8, 7),
          ).overlapsWith(other),
          isFalse,
        );
      },
    );
  });
}


lib/src/extensions/datetime_extension.dart
import 'package:flutter/material.dart';

/// This extension adds features to the [DateTime] class.
///
extension DateTimeExtension on DateTime {
  /// Returns whether this date **belongs** to the [dateTimeRange] date range
  /// *(including extremes)* or not ([dateTimeRange]`.start` ⩽ this ⩽
  /// [dateTimeRange]`.end`).
  ///
  /// Given the following:
  ///
  /// ```dart
  /// final DateTime start = DateTime(2022, 7);
  /// final DateTime end = DateTime(2022, 7, 31);
  /// final DateTimeRange range = DateTimeRange(start: start, end: end);
  /// ```
  ///
  /// The following examples apply:
  ///
  /// ```dart
  /// DateTime(2022, 6, 30).belongsTo(range); // isFalse
  /// DateTime(2022, 7).belongsTo(range); // isTrue
  /// DateTime(2022, 7, 15).belongsTo(range); // isTrue
  /// DateTime(2022, 7, 31).belongsTo(range); // isTrue
  /// DateTime(2022, 8).belongsTo(range); // isFalse
  /// ```
  ///
  bool belongsTo(final DateTimeRange dateTimeRange) {
    final bool isIncluded =
        isAfter(dateTimeRange.start) && isBefore(dateTimeRange.end);
    final bool coincidesWithStart = isAtSameMomentAs(dateTimeRange.start);
    final bool coincidesWithEnd = isAtSameMomentAs(dateTimeRange.end);

    return isIncluded || coincidesWithStart || coincidesWithEnd;
  }
}


lib/src/extensions/timeofday_extension.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';

/// This extension adds features to the [TimeOfDay] class.
///
extension TimeOfDayExtension on TimeOfDay {
  /// Returns if this time **is after** the [other] (this > [other]).
  ///
  /// Examples:
  ///
  /// ```dart
  /// const TimeOfDay(hour: 5, minute: 57)
  ///     .isAfter(const TimeOfDay(hour: 7, minute: 12)); // isFalse
  /// const TimeOfDay(hour: 7, minute: 35)
  ///     .isAfter(const TimeOfDay(hour: 7, minute: 57)); // isFalse
  /// const TimeOfDay(hour: 7, minute: 35)
  ///     .isAfter(const TimeOfDay(hour: 7, minute: 35)); // isFalse
  /// const TimeOfDay(hour: 7, minute: 35)
  ///     .isAfter(const TimeOfDay(hour: 7, minute: 34)); // isTrue
  /// ```
  ///
  bool isAfter(final TimeOfDay other) =>
      hour > other.hour || (hour == other.hour && minute > other.minute);

  /// Returns if this time **is before** the [other] (this < [other]).
  ///
  /// Examples:
  ///
  /// ```dart
  /// const TimeOfDay(hour: 7, minute: 35)
  ///     .isBefore(const TimeOfDay(hour: 7, minute: 57)); // isTrue
  /// const TimeOfDay(hour: 7, minute: 35)
  ///     .isBefore(const TimeOfDay(hour: 7, minute: 35)); // isFalse
  /// const TimeOfDay(hour: 7, minute: 35)
  ///     .isBefore(const TimeOfDay(hour: 7, minute: 34)); // isFalse
  /// const TimeOfDay(hour: 12, minute: 20)
  ///     .isBefore(const TimeOfDay(hour: 7, minute: 10)); // isFalse
  /// ```
  ///
  bool isBefore(final TimeOfDay other) =>
      hour < other.hour || (hour == other.hour && minute < other.minute);

  /// Returns the **string representation** of this time with:
  ///
  /// * the **[hour]s** and
  /// * the **[minute]s**.
  ///
  /// Examples:
  ///
  /// ```dart
  /// const TimeOfDay(hour: 7, minute: 15).toStringHm(); // '07:15'
  /// const TimeOfDay(hour: 15, minute: 5).toStringHm(); // '15:05'
  /// const TimeOfDay(hour: 12, minute: 35).toStringHm(); // '12:35'
  /// ```
  ///
  String toStringHm() {
    final DateTime date = DateTime(0, 0, 0, hour, minute);
    return DateFormat.Hm().format(date);
  }
}


lib/src/extensions/datetimerange_extension.dart
import 'package:flutter/material.dart';

import 'datetime_extension.dart';

/// This extension adds features to the [DateTimeRange] class.
///
extension DateTimeRangeExtension on DateTimeRange {
  /// Returns if this date range **belongs to** the [other] date range, that is,
  /// it is *completely included* in it ([other]`.start` ⩽ [start] ⩽
  /// [other]`.end`, and [other]`.start` ⩽ [end] ⩽ [other]`.end`).
  ///
  /// Given the following:
  ///
  /// ```dart
  /// final DateTimeRange other =
  ///     DateTimeRange(start: DateTime(2022), end: DateTime(2022, 7, 31));
  /// ```
  ///
  /// The following examples apply:
  ///
  /// ```dart
  /// DateTimeRange(
  ///   start: DateTime(2021, 9),
  ///   end: DateTime(2022),
  /// ).belongsTo(other); // isFalse
  /// DateTimeRange(
  ///   start: DateTime(2021, 9),
  ///   end: DateTime(2022, 1, 2),
  /// ).belongsTo(other); // isFalse
  /// DateTimeRange(
  ///   start: DateTime(2022),
  ///   end: DateTime(2022, 7, 31),
  /// ).belongsTo(other); // isTrue
  /// DateTimeRange(
  ///   start: DateTime(2022, 1, 2),
  ///   end: DateTime(2022, 7, 30),
  /// ).belongsTo(other); // isTrue
  /// DateTimeRange(
  ///   start: DateTime(2022, 7, 30),
  ///   end: DateTime(2022, 8),
  /// ).belongsTo(other); // isFalse
  /// ```
  ///
  bool belongsTo(final DateTimeRange other) {
    final bool startBelongsTo = start.belongsTo(other);
    final bool endBelongsTo = end.belongsTo(other);
    return startBelongsTo && endBelongsTo;
  }

  /// Returns if this date range **overlaps**, in part or completely, with the
  /// [other] date range ([other]`.start` ⩽ [start] ⩽ [other]`.end`, or
  /// [other]`.start` ⩽ [end] ⩽ [other]`.end`).
  ///
  /// Given the following:
  ///
  /// ```dart
  /// final DateTimeRange other =
  ///     DateTimeRange(start: DateTime(2022), end: DateTime(2022, 7, 31));
  /// ```
  ///
  /// The following examples apply:
  ///
  /// ```dart
  /// DateTimeRange(
  ///   start: DateTime(2021, 9),
  ///   end: DateTime(2021, 9, 5),
  /// ).overlapsWith(other); // isFalse
  /// DateTimeRange(
  ///   start: DateTime(2021, 9),
  ///   end: DateTime(2022),
  /// ).overlapsWith(other); // isTrue
  /// DateTimeRange(
  ///   start: DateTime(2021, 9),
  ///   end: DateTime(2022, 1, 2),
  /// ).overlapsWith(other); // isTrue
  /// DateTimeRange(
  ///   start: DateTime(2022),
  ///   end: DateTime(2022, 7, 31),
  /// ).overlapsWith(other); // isTrue
  /// DateTimeRange(
  ///   start: DateTime(2022, 1, 2),
  ///   end: DateTime(2022, 7, 30),
  /// ).overlapsWith(other); // isTrue
  /// DateTimeRange(
  ///   start: DateTime(2022, 7, 30),
  ///   end: DateTime(2022, 8),
  /// ).overlapsWith(other); // isTrue
  /// DateTimeRange(
  ///   start: DateTime(2022, 8),
  ///   end: DateTime(2022, 8, 7),
  /// ).overlapsWith(other); // isFalse
  /// ```
  ///
  bool overlapsWith(final DateTimeRange other) {
    final bool startBelongsTo = start.belongsTo(other);
    final bool endBelongsTo = end.belongsTo(other);
    return startBelongsTo || endBelongsTo;
  }
}


lib/src/extensions/list_datetimerange_extension.dart
import 'package:flutter/material.dart';

import 'datetimerange_extension.dart';

/// This extension adds features to the [List<DateTimeRange>] class.
///
extension ListDateTimeRangeExtension on List<DateTimeRange> {
  /// Returns the sum of the **duration** of *all* intervals in this list *after
  /// eliminating all duplicate intervals* (which overlap some interval already
  /// present in this list).
  ///
  /// To eliminate duplicate ranges this method uses the [minimize] method.
  ///
  /// Given the following:
  ///
  /// ```dart
  /// final DateTimeRange range1 =
  ///     DateTimeRange(start: DateTime(2022), end: DateTime(2022, 7));
  /// final DateTimeRange range2 = DateTimeRange(
  ///   start: DateTime(2022, 1, 25),
  ///   end: DateTime(2022, 4, 7),
  /// );
  /// final DateTimeRange range3 =
  ///     DateTimeRange(start: DateTime(2022, 4, 7), end: DateTime(2022, 7));
  /// final DateTimeRange range4 = DateTimeRange(
  ///   start: DateTime(2022, 5, 2),
  ///   end: DateTime(2022, 6, 15),
  /// );
  /// final DateTimeRange range5 = DateTimeRange(
  ///   start: DateTime(2022, 7, 3),
  ///   end: DateTime(2022, 9, 30),
  /// );
  /// final DateTimeRange range6 =
  ///     DateTimeRange(start: DateTime(2022), end: DateTime(2022, 7));
  /// final DateTimeRange range7 = range5;
  /// ```
  ///
  /// The following examples apply:
  ///
  /// ```dart
  /// <DateTimeRange>[range1, range2, range3, range4, range5]
  ///     .duration
  ///     .inDays; // 270
  /// <DateTimeRange>[range6, range7]
  ///     .duration
  ///     .inDays; // 270
  /// ```
  ///
  Duration get duration {
    final List<DateTimeRange> minimized = minimize();
    return minimized.fold<Duration>(
      Duration.zero,
      (final previousValue, final dateTimeRange) =>
          previousValue + dateTimeRange.duration,
    );
  }

  /// Returns a **list of date ranges** that **covers the entirety** of the
  /// period of the sum of all date ranges in this list, but *without repeating
  /// any overlap*.
  ///
  /// Given the following:
  ///
  /// ```dart
  /// final DateTimeRange range1 =
  ///     DateTimeRange(start: DateTime(2022), end: DateTime(2022, 7));
  /// final DateTimeRange range2 = DateTimeRange(
  ///   start: DateTime(2022, 1, 25),
  ///   end: DateTime(2022, 4, 7),
  /// );
  /// final DateTimeRange range3 =
  ///     DateTimeRange(start: DateTime(2022, 4, 7), end: DateTime(2022, 7));
  /// final DateTimeRange range4 = DateTimeRange(
  ///   start: DateTime(2022, 5, 2),
  ///   end: DateTime(2022, 6, 15),
  /// );
  /// final DateTimeRange range5 = DateTimeRange(
  ///   start: DateTime(2022, 7, 3),
  ///   end: DateTime(2022, 9, 30),
  /// );
  /// final DateTimeRange range6 =
  ///     DateTimeRange(start: DateTime(2022), end: DateTime(2022, 7));
  /// final DateTimeRange range7 = range5;
  /// ```
  ///
  /// The following examples apply:
  ///
  /// ```dart
  /// <DateTimeRange>[range1, range2, range3, range4, range5].minimize();
  ///   // <DateTimeRange>[range6, range7]
  /// <DateTimeRange>[range1, range2, range3].minimize();
  ///   // <DateTimeRange>[range6]
  /// ```
  ///
  List<DateTimeRange> minimize() {
    final List<DateTimeRange> result = [];
    List<DateTimeRange> resultTemp = [];

    List<DateTimeRange> minimizeInternal(final List<DateTimeRange> initial) {
      bool hasAnythingBeenRemoved = false;
      initial.sort((final r1, final r2) {
        // 1º comparison
        final int comparison1 = r1.start.compareTo(r2.start);
        if (comparison1 != 0) return comparison1;
        // Last comparison
        final int comparison2 = r1.end.compareTo(r2.end);
        return comparison2;
      });
      result.add(initial.first);
      for (int i = 1; i < initial.length; i++) {
        if (initial[i].overlapsWith(result.last)) {
          final List<DateTime> fourDates = [
            result.last.start,
            result.last.end,
            initial[i].start,
            initial[i].end,
          ]..sort();
          final DateTimeRange joinRange = DateTimeRange(
            start: fourDates.first,
            end: fourDates.last,
          );
          result
            ..removeLast()
            ..add(joinRange);
          hasAnythingBeenRemoved = true;
        } else {
          result.add(initial[i]);
        }
      }
      if (hasAnythingBeenRemoved) {
        resultTemp = [...result];
        result.clear();
        minimizeInternal(resultTemp);
      }
      return result;
    }

    minimizeInternal(this);
    return result;
  }
}


example/helpers_plethora_flutter_example.dart
import 'package:flutter/material.dart';
import 'package:helpers_plethora_flutter/helpers_plethora_flutter.dart';

void main() {
  /// DateTimeExtension
  final DateTime start = DateTime(2022, 7);
  final DateTime end = DateTime(2022, 7, 31);
  final DateTimeRange range = DateTimeRange(start: start, end: end);
  DateTime(2022, 6, 30).belongsTo(range);

  /// DateTimeRangeExtension
  final DateTimeRange other =
      DateTimeRange(start: DateTime(2022), end: DateTime(2022, 7, 31));
  DateTimeRange(
    start: DateTime(2021, 9),
    end: DateTime(2022),
  ).belongsTo(other);
  DateTimeRange(
    start: DateTime(2022, 8),
    end: DateTime(2022, 8, 7),
  ).overlapsWith(other);

  /// ListDateTimeRangeExtension
  final DateTimeRange range1 =
      DateTimeRange(start: DateTime(2022), end: DateTime(2022, 7));
  final DateTimeRange range2 = DateTimeRange(
    start: DateTime(2022, 1, 25),
    end: DateTime(2022, 4, 7),
  );
  final DateTimeRange range3 =
      DateTimeRange(start: DateTime(2022, 4, 7), end: DateTime(2022, 7));
  final DateTimeRange range4 = DateTimeRange(
    start: DateTime(2022, 5, 2),
    end: DateTime(2022, 6, 15),
  );
  final DateTimeRange range5 = DateTimeRange(
    start: DateTime(2022, 7, 3),
    end: DateTime(2022, 9, 30),
  );
  <DateTimeRange>[range1, range2, range3, range4, range5].duration.inDays;
  <DateTimeRange>[range1, range2, range3, range4, range5].minimize();

  /// TimeOfDayExtension
  const TimeOfDay(hour: 5, minute: 57)
      .isAfter(const TimeOfDay(hour: 7, minute: 12));
  const TimeOfDay(hour: 7, minute: 35)
      .isBefore(const TimeOfDay(hour: 7, minute: 34));
  const TimeOfDay(hour: 15, minute: 5).toStringHm();
}
